### DISCUSSION TITLE
Reusing request extractor without consuming the extensions

### DISCUSSION THREAD

**Discussion by tony707** [tokio-rs/axum#1899] [Q&A]
Hi,

In my quest to reuse existing body extractors such as  `Bytes` and `Json`, I'd like to avoid consuming all the request extensions and retrieve only the ones axum need to actually do the extraction. In the following case I'd like to get and clone the `DefaultBodyLimitKind` extension for example, but it's private to the crate. Would you consider making it public ? Or do you have any alternative option to suggest ?

```rust
#[async_trait]
impl<S, B> FromRequest<S, B> for MyBytesExtractor
where
    B: HttpBody + Send + 'static,
    B::Data: Send,
    B::Error: Into<BoxError>,
    S: Send + Sync,
{
    type Rejection = BytesRejection;

    async fn from_request(req: Request<B>, state: &S) -> Result<Self, Self::Rejection> {
        let (parts, body) = req.into_parts();

        let (mut inner_parts, body) = Request::new(body).into_parts();
        inner_parts.headers = parts.headers.clone();
        // Insert parts.extensions.get<DefaultBodyLimitKind>() ?.
        let inner_request = Request::from_parts(inner_parts, body);

        let bytes = Bytes::from_request(inner_request, state).await?;

        Ok(Self(parts, bytes))
    }
}
```

Thanks!

**Comment by jplatte** [comment]
Our own `Bytes` `FromRequest` impl no longer extracts `DefaultBodyLimitKind` directly, have you checked the current implementation? Maybe that will help you in implementing your own.

**Reply by tony707** [reply]
Yes I've looked at it, 

In fact what I wished is that the `FromRequest` on `Bytes` could have returned the remaining request parts in some way instead of loosing them inside the call to `into_limited_body()`. That could even be applied to all body extractors ?

In the end I've implemented my own, let me know if that is changing in any way in future versions.

Thanks!

**Reply by davidpdrsn** [reply]
> In the end I've implemented my own

I think that's fine. 

Not sure it's worth changing the traits since it would mostly be for convenience.
